keps/sig-network/1880-services-ip-ranges/README.md

```diff
@@ -0,0 +1,885 @@
+
+For keeping backwards compatibility, an special IP Range has to match the ones created by the kube-apiserver flags. 
+This special IP Range also defines the primary IP family of the Cluster.
+
+### User Stories
+
+#### Story 1
+
+As a Kubernetes user I want to be able to create and use multiple Service Subnets.
+
+#### Story 2
+
+As a Kubernetes admin I want to be able to resize my current Service subnet with zero downtime, or
+requiring complex and "unsupported" operations.
+
+#### Story 3
+
+As a Kubernetes developer I want to be able to know the current Service IP range and the IP addresses allocated.
+
+#### Story 4
+
+As a Kubernetes admin that wants to use IPv6, I want to be able to follow the IPv6 address
+planning best practices, being able to assign /64 prefixes to my end subnets.
+
+#### Story 5
+
+As a Kubernetes admin using a dual-stack cluster I want to be able to change the primary IP family of the cluster.
+
+### Notes/Constraints/Caveats (Optional)
+
+Current API machinery doesn't consider transactions, Services API is kind of special in this regard since already
+performs allocations inside the Registry pipeline using the Storage to keep consistency on an Obscure API Object that
+stores the bitmap with the allocations.
+
+This proposal also requires to perform validation from one Object against other Object inside an API request.
+
+### Risks and Mitigations
+
+Cross validation of Objects is not common in the Kubernetes API, sig-apimachinery should verify not bad precedent is introduced.
+
+Current Kubernetes cluster have a very static network configuration, exposing the Service IP range configuration will give
+more flexibility to users, with the risk to have problematic or incongruent network configurations.
+
+The Pod Network configuration is not really a Cluster configuration option, some CNI plugins use the node.Spec.ClusterCIDR field,
+but that is not guaranteed to be used. There is an effort in keps/sig-network/2593-multiple-cluster-cidrs to expose this configuration,
+but this also presents other cross-validation problem.
+
+## Design Details
+
+### API
+
+Define two new API objects
+
+#### IPRange
+
+##### Type
+
+```go
+// IPRange defines a range of IPs using CIDR format (192.168.0.0/24 or 2001:db2::0/64).
+type IPRange struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	Spec   IPRangeSpec   `json:"spec,omitempty"`
+	Status IPRangeStatus `json:"status,omitempty"`
+}
+
+
+// IPRangeSpec describe how the IPRange's specification looks like.
+type IPRangeSpec struct {
+	// An IPv4 block in CIDR notation "10.0.0.0/8"
+	IPv4 string `json:"ipv4"`
+
+	// An IPv6 block in CIDR notation "fd12:3456:789a:1::/64"
+	IPv6 string `json:"ipv6"`
+
+  // Users may reserve the first number of IPs in each range to be not used for automatic allocation.
+  ReservedSize int
+
+  // Primary defines the primary IP family of the IP range (default to IPv4)
+  Primary enum(ipv4|ipv6)
```

Comment:
Adding a "Primary family" here seems strange, even if we decide that ranges should be dual-family. Instead, that seems like an aspect of the allocator itself.

What if there are two ranges, and one has v4 primary and the other v6? What is the default action?

If this sort of live configuration is desired, then some kind of ServiceIPPolicy object might be a better fit.

Reply:
this is for backwards compatibility, we already have this notion of "primary" IPFamily, and the whole defaulting and policy logic is very complex.

The Primary only defines the defaulting, one service can only reference one IPRange, that is other reason why IPRange contains both families ...

Reply:
That seems unnecessarily specific. Rather, Services could request a specific ServiceIPClass (to use the existing Kubernetes practices), with a certain ServiceIPClass being the default (same as Ingress or PVs or Runtimes).

Because if the cluster administrator wants to "silently" enable disjoint ranges for whatever reason, they should be able to do this without impacting users.

(Perhaps this needs to be broken out as a separate thread).

Reply:
> That seems unnecessarily specific. 

One of the use cases is to be able to switch the primary IPFamily, the API Service allocation logic depends on the primary IP family ... I still didn't work out all the details here, but this notion of primary IP family is something we should be able to handle ... let me write up the allocation logic, so I can refresh my memory and we can discuss there in more detail

> That seems unnecessarily specific. Rather, Services could request a specific ServiceIPClass (to use the existing Kubernetes practices), with a certain ServiceIPClass being the default (same as Ingress or PVs or Runtimes).

> Because if the cluster administrator wants to "silently" enable disjoint ranges for whatever reason, they should be able to do this without impacting users.
> (Perhaps this needs to be broken out as a separate thread).

Yes, this is an important topic, which model do we want to implement, a Class model or as @thockin mentioned yesterday in sig-network meeting, if I understood him correctly is, where IPs are fungible resources, and IPRanges are blocks of these resources, you just simple add more pools to add more IPs resources and the apiserver keeps consuming them.

Reply:
removed the primary field